contracts/UrlContract.sol

```diff
@@ -4,117 +4,92 @@ pragma solidity ^0.8.9;
-        );
-
-        // Create a new URL struct and store it in the mapping
-        urlToMetadata[_url] = URL(_url, 0);
-
-        // Add the URL to the array of all submitted URLs
-        urlArray.push(urlToMetadata[_url]);
+    
+    struct User {
+        uint256[] upvotedURLs;
+        uint256[] downvotedURLs;
+        uint256[] submittedURLs;
     }
+    
+    URL[] public urls;
+    Tag[] public tags;
+    mapping (address => User) internal users;
 
-    // Function to upvote a URL
-    function upvoteURL(string memory _url) public {
-        // Increment the vote count for the URL
-        urlToMetadata[_url].voteCount++;
 
-        // Re-sort the array of URLs by vote count
-        sortURLArray();
+    function getUpvotedURLs(address userAddress) public view returns (uint256[] memory) {
+        return users[userAddress].upvotedURLs;
     }
 
-    // function to return upvote/downvote count of a URL
-    function getcount(string memory _url) public view returns (int) {
-        // return votecount
-        return urlToMetadata[_url].voteCount;
+    function getDownvotedURLs(address userAddress) public view returns (uint256[] memory) {
+        return users[userAddress].downvotedURLs;
     }
 
-    // Function to downvote a URL
-    function downvoteURL(string memory _url) public {
-        // Decrement the vote count for the URL
-        urlToMetadata[_url].voteCount--;
-
-        // Re-sort the array of URLs by vote count
-        sortURLArray();
+    function getSubmittedURLs(address userAddress) public view returns (uint256[] memory) {
+        return users[userAddress].submittedURLs;
     }
 
-    // Helper function to sort the array of URLs by vote count
-    function sortURLArray() private {
-        // Sort the array in place using a bubble sort algorithm
-        for (uint i = 0; i < urlArray.length; i++) {
-            for (uint j = i + 1; j < urlArray.length; j++) {
-                if (urlArray[i].voteCount < urlArray[j].voteCount) {
-                    // Swap the URLs
-                    URL memory temp = urlArray[i];
-                    urlArray[i] = urlArray[j];
-                    urlArray[j] = temp;
-                }
-            }
-        }
+    function submitURL(string memory title, string memory url, uint256[] memory tagIds) public {
+    require(bytes(title).length > 0 && bytes(url).length > 0 && tagIds.length > 0, "Invalid input");
+    for (uint256 i = 0; i < tagIds.length; i++) {
+        require(tagIds[i] < tags.length, "Invalid tag id");
     }
-
-    // Function to return the top 10 URLs
-    function getTopURLs() public view returns (string[] memory) {
-        // Return a slice of the array with the top 10 URLs
-        string[] memory top10 = new string[](urlArray.length);
-        for (uint i = 0; i < 10; i++) {
-            top10[i] = urlArray[i].url;
-        }
-        return top10;
+    URL memory newURL = URL(title, url, msg.sender, 0, 0, tagIds);
```

Comment:
msg.sender will always be same, since we are going to write to smart contract from our central hub server. so , I will need to pass another parameter here i think that will have user's wallet address. 

but the question remains, should i store such info ?